delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala

```diff
@@ -422,40 +422,57 @@ object Resources {
     def update(u: UpdateResource) = {
       import u.jsonld._
       // format: off
-      for {
-        s                          <- stateWhereResourceIsEditable(u)
-        schemaRef                   = u.schemaOpt.getOrElse(ResourceRef.Latest(s.schema.iri))
-        (schemaRev, schemaProject) <- validate(u.id, expanded, schemaRef, s.project, u.caller)
-        time                       <- clock.realTimeInstant
-      } yield ResourceUpdated(u.id, u.project, schemaRev, schemaProject, types, u.source, compacted, expanded, remoteContextRefs, s.rev + 1, time, u.subject, u.tag)
+      stateWhereResourceIsEditable(u).flatMap { s =>
+        val changesDetected = s.remoteContexts != remoteContextRefs || s.source != u.source
+        if(changesDetected) {
+          val schemaRef = u.schemaOpt.getOrElse(ResourceRef.Latest(s.schema.iri))
+          for {
+            (schemaRev, schemaProject) <- validate(u.id, expanded, schemaRef, s.project, u.caller)
+            time <- clock.realTimeInstant
+          } yield ResourceUpdated(u.id, u.project, schemaRev, schemaProject, types, u.source, compacted, expanded, remoteContextRefs, s.rev + 1, time, u.subject, u.tag)
+        } else {
+          // If no changes have been detected on the resource itself, we fallback on other commands or raise an error
+          (u.schemaOpt, u.tag) match {
+            case (Some(schema), tagOpt) => updateResourceSchema(UpdateResourceSchema(u.id, u.project, schema, u.rev, u.caller, tagOpt))
+            case (None, Some(tagValue)) => tag(TagResource(u.id, u.project, u.schemaOpt, s.rev, tagValue, u.rev, u.subject))
+            case (None, None) => IO.raiseError(NoChangeDetected(s))
+          }
+        }
+      }
       // format: on
     }
 
     def updateResourceSchema(u: UpdateResourceSchema) = {
       for {
-        s                          <- stateWhereResourceIsEditable(u)
-        (schemaRev, schemaProject) <- validate(u.id, u.expanded, u.schemaRef, s.project, u.caller)
-        types                       = u.expanded.getTypes.getOrElse(Set.empty)
+        s <- stateWhereResourceIsEditable(u)
+
+        (schemaRev, schemaProject) <- validate(u.id, s.expanded, u.schemaRef, s.project, u.caller)
+        _                          <- IO.raiseWhen(s.schema.iri == u.schemaRef.iri)(NoChangeDetected(s))
```

Comment:
For the update schema case, I think you are right even if it seems there is a small bug as in this implementation as you can't update a schema to use another tag of the same schema for example.
So I will remove it for now because we need to also save in the resource the schema (latest/tag/revision) requested by the user besides the revision of the schema the resource has been validated with.

For the refresh case, we want to raise an error if the resource is validated by the latest version of a schema as this one may have changed since the last update.